### ISSUE TITLE
Analytics > Orders reports aren't sent due to timing of actions

### ISSUE THREAD

**Issue by n-dawson** [woothemes/woocommerce#36755]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When we attempt to download larger reports from the Analytics section, there's a notice shown stating the report will be emailed however that email is never sent.

Although the `email_report_download_link` action is supposed to be dependent on `export_report`, what we're seeing is that action frequently completing before the export is ready resulting in the email failing to send.

We've stepped through it in Xdebug and found the download link action fires a single time with percentage complete being 0.

Initially we thought it could be due to using a real CRON task but the same behaviour occurs with WP_CRON enabled.

![Screenshot 2023-02-03 at 20 26 15](https://user-images.githubusercontent.com/3666437/216707723-21361d86-8811-455f-875b-7d02c500e21f.png)

![Screenshot 2023-02-03 at 20 48 01](https://user-images.githubusercontent.com/3666437/216707756-e9a13eb9-c1a1-4904-b448-8c49b67eb792.png)

### Expected behavior

The 'email_report_download_link' action should only run when the 'export_report' task is complete.

### Actual behavior

Action scheduler ignores the dependency and runs the action regardless. Whether the email sends is the result of whichever task completes first.

### Steps to reproduce

1. Go to Analytics -> Orders
2. Select "Last year" for the date range (our test site had 33 orders for that period)
3. Press Download
4. Go to WooCommerce -> Status -> Scheduled Actions to monitor the tasks

Use Mailhog or an email logging plugin to determine whether an email was sent.

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://reporting.test
Site address (URL): https://reporting.test
WC Version: 7.3.0
REST API Version: ✔ 7.3.0
WC Blocks Version: ✔ 9.1.5
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 7.4.30
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 7.86.0
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.3.0
WC Database Prefix: wp_
Total Database Size: 222.96MB
Database Data Size: 131.37MB
Database Index Size: 91.59MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_api_keys: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_woocommerce_attribute_taxonomies: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_downloadable_product_permissions: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_order_items: Data: 1.43MB + Index: 0.92MB + Engine MyISAM
wp_woocommerce_order_itemmeta: Data: 19.00MB + Index: 7.49MB + Engine MyISAM
wp_woocommerce_tax_rates: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_woocommerce_tax_rate_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zones: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zone_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zone_methods: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_payment_tokens: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_payment_tokenmeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
_mig_wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.06MB + Index: 0.05MB + Engine MyISAM
wp_actionscheduler_claims: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_actionscheduler_groups: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_actionscheduler_logs: Data: 0.04MB + Index: 0.04MB + Engine MyISAM
wp_automatewoo_abandoned_carts: Data: 0.10MB + Index: 0.01MB + Engine MyISAM
wp_automatewoo_customer_meta: Data: 0.01MB + Index: 0.02MB + Engine MyISAM
wp_automatewoo_customers: Data: 0.28MB + Index: 0.48MB + Engine MyISAM
wp_automatewoo_events: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_automatewoo_guest_meta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_automatewoo_guests: Data: 0.21MB + Index: 0.30MB + Engine MyISAM
wp_automatewoo_log_meta: Data: 0.85MB + Index: 0.48MB + Engine MyISAM
wp_automatewoo_logs: Data: 0.14MB + Index: 0.25MB + Engine MyISAM
wp_automatewoo_queue: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_automatewoo_queue_meta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_commentmeta: Data: 0.45MB + Index: 0.52MB + Engine InnoDB
wp_comments: Data: 8.52MB + Index: 9.09MB + Engine InnoDB
wp_ee_product_sync_call: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_ee_product_sync_data: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_ee_products_sync_list: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_ewwwio_images: Data: 0.41MB + Index: 0.50MB + Engine InnoDB
wp_gf_addon_feed: Data: 0.01MB + Index: 0.00MB + Engine MyISAM
wp_gf_draft_submissions: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_gf_entry: Data: 2.12MB + Index: 0.22MB + Engine MyISAM
wp_gf_entry_meta: Data: 6.75MB + Index: 4.53MB + Engine MyISAM
wp_gf_entry_notes: Data: 1.21MB + Index: 0.32MB + Engine MyISAM
wp_gf_form: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_gf_form_meta: Data: 0.26MB + Index: 0.00MB + Engine MyISAM
wp_gf_form_revisions: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_gf_form_view: Data: 1.25MB + Index: 0.55MB + Engine MyISAM
wp_gf_rest_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_groups_capability: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_groups_group: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_groups_group_capability: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_groups_user_capability: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_groups_user_group: Data: 0.11MB + Index: 0.08MB + Engine InnoDB
wp_itsec_bans: Data: 0.09MB + Index: 0.13MB + Engine InnoDB
wp_itsec_dashboard_events: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_itsec_distributed_storage: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_itsec_fingerprints: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_itsec_geolocation_cache: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_itsec_lockouts: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_itsec_logs: Data: 0.07MB + Index: 0.01MB + Engine MyISAM
wp_itsec_mutexes: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_itsec_opaque_tokens: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_itsec_temp: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_itsec_user_groups: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mc4wp_log: Data: 0.96MB + Index: 0.06MB + Engine MyISAM
wp_options: Data: 13.34MB + Index: 0.28MB + Engine InnoDB
wp_pmxi_files: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_pmxi_hash: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_pmxi_history: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_pmxi_images: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_pmxi_imports: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_pmxi_posts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_pmxi_templates: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_postmeta: Data: 38.58MB + Index: 38.16MB + Engine InnoDB
wp_posts: Data: 8.52MB + Index: 3.64MB + Engine InnoDB
wp_redirection_404: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_redirection_groups: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_redirection_items: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_redirection_logs: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_rg_constantcontact: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_rg_form: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_rg_form_meta: Data: 0.13MB + Index: 0.00MB + Engine InnoDB
wp_rg_form_view: Data: 1.52MB + Index: 0.53MB + Engine InnoDB
wp_rg_incomplete_submissions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_rg_lead: Data: 1.52MB + Index: 0.19MB + Engine InnoDB
wp_rg_lead_detail: Data: 3.52MB + Index: 9.06MB + Engine InnoDB
wp_rg_lead_detail_long: Data: 0.44MB + Index: 0.00MB + Engine InnoDB
wp_rg_lead_meta: Data: 0.14MB + Index: 0.22MB + Engine InnoDB
wp_rg_lead_notes: Data: 0.09MB + Index: 0.03MB + Engine InnoDB
wp_signups: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.08MB + Index: 0.05MB + Engine InnoDB
wp_term_taxonomy: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.02MB + Engine MyISAM
wp_terms: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 7.52MB + Index: 9.03MB + Engine InnoDB
wp_users: Data: 0.36MB + Index: 0.33MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.01MB + Engine MyISAM
wp_wc_admin_notes: Data: 0.04MB + Index: 0.00MB + Engine MyISAM
wp_wc_category_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_customer_lookup: Data: 0.40MB + Index: 0.28MB + Engine MyISAM
wp_wc_download_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_order_bundle_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.13MB + Index: 0.24MB + Engine MyISAM
wp_wc_order_product_lookup: Data: 1.86MB + Index: 1.36MB + Engine MyISAM
wp_wc_order_stats: Data: 0.49MB + Index: 0.34MB + Engine MyISAM
wp_wc_order_tax_lookup: Data: 0.04MB + Index: 0.05MB + Engine MyISAM
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.01MB + Index: 0.02MB + Engine MyISAM
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_tax_rate_classes: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_webhooks: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wcsp_global_discount_rule_mapping: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wcsp_group_category_pricing_mapping: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wcsp_role_category_pricing_mapping: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wcsp_user_category_pricing_mapping: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_woocommerce_bundled_itemmeta: Data: 0.09MB + Index: 0.14MB + Engine InnoDB
wp_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wpml_mails: Data: 6.52MB + Index: 0.00MB + Engine InnoDB
wp_wpmm_subscribers: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wprm_analytics: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wprm_ratings: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wt_iew_action_history: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wt_iew_cron: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wt_iew_ftp: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wt_iew_mapping_template: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wusp_group_product_price_mapping: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wusp_role_pricing_mapping: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wusp_rules: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wusp_subrules: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_wusp_user_pricing_mapping: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_yoast_indexable: Data: 0.30MB + Index: 0.06MB + Engine MyISAM
wp_yoast_indexable_hierarchy: Data: 0.05MB + Index: 0.13MB + Engine MyISAM
wp_yoast_migrations: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_yoast_primary_term: Data: 0.01MB + Index: 0.01MB + Engine MyISAM
wp_yoast_seo_links: Data: 0.05MB + Index: 0.02MB + Engine MyISAM
wp_yoast_seo_meta: Data: 0.01MB + Index: 0.01MB + Engine MyISAM

### Post Type Counts ###

acf-field: 356
acf-field-group: 37
amn_mi-lite: 2
attachment: 1206
aw_workflow: 4
itsec-dash-card: 5
itsec-dashboard: 1
mc4wp-form: 1
nav_menu_item: 86
oembed_cache: 10
page: 34
post: 147
product: 45
product_variation: 137
recipe: 39
revision: 1453
shop_bogof_rule: 1
shop_coupon: 22
shop_order: 7270
shop_order_refund: 111
wp_block: 1
wp_global_styles: 1
wp_log: 212
wprm_recipe: 39

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (23) ###

Gravity Forms: by Gravity Forms – 2.6.7
Advanced Custom Fields PRO: by Delicious Brains – 5.12.3
iThemes Security: by iThemes – 8.1.4
CheckoutWC: by Objectiv – 7.8.7
Duplicate Page: by mndpsingh287 – 4.5
Gravity Forms Address Autocomplete: by wpexperts.io – 1.7
Gravity Forms Address Enhanced: by WebAware – 1.25.0
WP Gravity Forms Salesforce: by CRM Perks. – 1.3.6
Gravity Forms Mailchimp Add-On: by Gravity Forms – 5.1
Gravity Forms User Registration Add-On: by Gravity Forms – 5.0
Gutenberg: by Gutenberg Team – 14.7.3
Intuitive Custom Post Order: by hijiri – 3.1.3
MC4WP: Mailchimp for WordPress: by ibericode – 4.8.12
MC4WP: Mailchimp for WordPress Premium: by ibericode – 4.8
Redirection: by John Godley – 5.3.6
Safe SVG: by 10up – 2.0.3
User Role Editor: by Vladimir Garagulya – 4.63.2
WooCommerce Stripe Gateway: by WooCommerce – 7.0.1
WooCommerce Product Reviews Pro: by SkyVerge – 1.16.1 (update to version 1.18.0 is available)
WooCommerce Shipment Tracking: by WooCommerce – 1.8.0 (update to version 2.0.0 is available)
WooCommerce: by Automattic – 7.3.0
Yoast SEO: by Team Yoast – 20.0
WP Recipe Maker: by Bootstrapped Ventures – 8.6.2

### Must Use Plugins (1) ###

custom-entities.php: by  –

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: bundle (bundle)
external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: –
Order datastore: WC_Order_Data_Store_CPT

### WC Pages ###

Shop base: #1976 - /shop/
Cart: #1977 - /cart/
Checkout: #1978 - /checkout/
My account: #1979 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty
Version: 2.1
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2023-02-04 13:35:43 -05:00
Options: ✔
Notes: 99
Onboarding: completed

### Action Scheduler ###

Complete: 168
Oldest: 2023-01-03 16:00:01 -0500
Newest: 2023-02-03 15:26:08 -0500

Failed: 17
Oldest: 2022-08-14 10:59:51 -0400
Newest: 2022-09-01 15:06:10 -0400

Pending: 2
Oldest: 2023-02-03 18:39:40 -0500
Newest: 2023-02-10 07:06:31 -0500

### Status report information ###

Generated at: 2023-02-03 15:53:04 -05:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by coreymckrill** [comment]
If I'm understanding the issue correctly, this probably isn't actually a problem in Action Scheduler. Action Scheduler itself does not support a concept of actions being dependent on other actions, so the root of the problem is likely in the analytics code.

**Comment by n-dawson** [comment]
@coreymckrill I didn't look too closely into how the Action Scheduler works so it's entirely possible I've misunderstood what's going on. There's a method on the ReportExporter class which lead me to believe the email task would be dependent on the report exporting.

```
public static function get_dependencies() {
	return array(
		'email_report_download_link' => self::get_action( 'export_report' ),
	);
}
```

**Comment by Saggre** [comment]
@n-dawson Could you check if this fixes your problem? Can you replicate this 100% of the time?

**Comment by mikkamp** [comment]
The WooCommerce 9.3 release included PR #50082 

The main idea of that PR was to ensure that the link is only sent after the report has been completed, and reschedule the action if it does happen to run earlier. Based on that information I'm considering the scenario in this issue as resolved.

Feel free to let us know if there is still any remaining issue with the latest version of WooCommerce and we can have another look to see if there is anything else that could still trigger this to run incorrectly.
